File: completed/2657.swift
class Solution {
    func findThePrefixCommonArray(_ A: [Int], _ B: [Int]) -> [Int] {
        var aSet = Set<Int>()
        var bSet = Set<Int>()
        var count = 0
        return zip(A, B).map { (a, b) in
            aSet.insert(a)
            bSet.insert(b)
            if aSet.contains(b) {
                count += 1
                aSet.remove(b)
                bSet.remove(b)
            }
            if bSet.contains(a) {
                count += 1
                aSet.remove(a)
                bSet.remove(a)
            }
            return count
        }
    }
}

File: completed/2657.java
class Solution {
    public int[] findThePrefixCommonArray(int[] A, int[] B) {
        int n = A.length;
        boolean[] aCounted = new boolean[n];
        boolean[] bCounted = new boolean[n];
        int[] prefixCommonArray = new int[n];
        for (int i = 0; i < n; i++) {
            if (i > 0) {
                prefixCommonArray[i] = prefixCommonArray[i - 1];
            }
            aCounted[A[i] - 1] = true;
            bCounted[B[i] - 1] = true;
            if (A[i] == B[i]) {
                prefixCommonArray[i]++;
            } else {
                if (aCounted[A[i] - 1] && bCounted[A[i] - 1]) {
                    prefixCommonArray[i]++;
                }
                if (aCounted[B[i] - 1] && bCounted[B[i] - 1]) {
                    prefixCommonArray[i]++;
                }
            }
        }
        return prefixCommonArray;
    }
}
